x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java

```diff
@@ -172,71 +168,38 @@ public void putAsync(ILMHistoryItem item) {
      */
-    @SuppressWarnings("unchecked")
-    static void ensureHistoryIndex(Client client, ClusterState state, ActionListener<Boolean> listener) {
-        final String initialHistoryIndexName = ILM_HISTORY_INDEX_PREFIX + "000001";
-        final IndexAbstraction ilmHistory = state.metadata().getIndicesLookup().get(ILM_HISTORY_ALIAS);
-        final IndexAbstraction initialHistoryIndex = state.metadata().getIndicesLookup().get(initialHistoryIndexName);
-
-        if (ilmHistory == null && initialHistoryIndex == null) {
-            // No alias or index exists with the expected names, so create the index with appropriate alias
-            logger.debug("creating ILM history index [{}]", initialHistoryIndexName);
-
-            // Template below should be already defined as real index template but it can be deleted. To avoid race condition with its
-            // recreation we apply settings and mappings ourselves
-            byte[] templateBytes = TEMPLATE_ILM_HISTORY.loadBytes();
-            Map<String, Object> templateAsMap = XContentHelper.convertToMap(new BytesArray(templateBytes, 0, templateBytes.length),
-                false, XContentType.JSON).v2();
-
-            client.admin().indices().prepareCreate(initialHistoryIndexName)
-                .setSettings((Map<String, ?>) templateAsMap.get("settings"))
-                .setMapping((Map<String, Object>) templateAsMap.get("mappings"))
-                .setWaitForActiveShards(1)
-                .addAlias(new Alias(ILM_HISTORY_ALIAS).writeIndex(true).isHidden(true))
-                .execute(new ActionListener<>() {
+    static void ensureHistoryDataStream(Client client, ClusterState state, ActionListener<Boolean> listener) {
+        if (state.getMetadata().dataStreams().containsKey(ILM_HISTORY_ALIAS)) {
+            listener.onResponse(false);
+        } else {
+            logger.debug("creating ILM history data stream [{}]", ILM_HISTORY_ALIAS);
+            client.execute(CreateDataStreamAction.INSTANCE, new CreateDataStreamAction.Request(ILM_HISTORY_ALIAS),
+                new ActionListener<AcknowledgedResponse>() {
                     @Override
-                    public void onResponse(CreateIndexResponse response) {
+                    public void onResponse(AcknowledgedResponse acknowledgedResponse) {
                         listener.onResponse(true);
+
                     }
 
                     @Override
                     public void onFailure(Exception e) {
-                        if (e instanceof ResourceAlreadyExistsException) {
-                            // The index didn't exist before we made the call, there was probably a race - just ignore this
-                            logger.debug("index [{}] was created after checking for its existence, likely due to a concurrent call",
-                                initialHistoryIndexName);
+                        if(e instanceof IllegalArgumentException && e.getMessage().contains("exists")){
```

Comment:
should we maybe probe the error message to be an exact match? ie. `data_stream [ilm-history-3] already exists` (with corresponding parameterized  data stream name)
I wonder why we're not throwing `ResourceAlreadyExists` in the data stream case too.

Reply:
Changed this to throw `ResourceAlreadyExistsException` in https://github.com/elastic/elasticsearch/pull/60518